### PR TITLE
Dev 0.1

### PR DESCRIPTION
**Pull request by wli51**
Merging a working buddi_v2 dev-0.1 branch that has examples into main

### CODE REVIEW

File: examples/3.train_buddi4.ipynb [hunk 1]
```diff
@@ -0,0 +1,933 @@
+{
```
Comment (nrosed) [inline examples/3.train_buddi4.ipynb:1]: <p>much prettier!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NDavidsonLab/buddi_v2/pull/1/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

File: src/buddi_v2/models/buddi4.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    X_hat_supervised = supervised_decoder(supervised_decoder_concat)
+    X_hat_unsupervised = unsupervised_decoder(unsupervised_decoder_concat)
+
+    # --------------------- Losses ---------------------
+    ## Note these are all functions with standard tensorflow signature accepting y_true and y_pred
+    kl_loss_fn_label = kl_loss_generator(
+        beta=1.0,#beta_kl_label, 
+        agg_fn=K.sum, axis=-1)
+    kl_loss_fn_stim = kl_loss_generator(
+        beta=1.0,#beta_kl_stim,
```
Comment (nrosed) [inline src/buddi_v2/models/buddi4.py:226]: I'm probably missing something, but why is it set to 1?

File: src/buddi_v2/plotting/plot_latent_space.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
+import seaborn as sns
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+
+from sklearn.decomposition import PCA
+from sklearn.manifold import TSNE
+import umap.umap_ as umap
+
+def _get_projection(z, type='PCA'):
```
Comment (nrosed) [inline src/buddi_v2/plotting/plot_latent_space.py:10]: nice!

File: examples/1.prepare_pseudobulk.ipynb [hunk 4]
```diff
@@ -0,0 +1,478 @@
+{
```
[the remark this answers is not in the document]
Comment (wli51) [reply examples/1.prepare_pseudobulk.ipynb:1]: <p>Yes absolutely!</p>

File: examples/3.train_buddi4.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (wli51) [reply examples/3.train_buddi4.ipynb:1]: <p>My primary motivation for not using MAE  (or any other loss function different from the other classifiers) for the proportion estimator is to have all classifier loss be on the same scale to so the alpha weights can better control the aspect we want buddi to focus on (although this example does not specify different classifier weights). In the past I find CCEs to be a stronger loss function than MAE at enforcing predictions of soft-maxed vectors to be close to the target, especially in the context of VAEs with multiple predictions. The second reasoning I had was that in context of deconvolution, the same absolute error can have different importance. a 0.05 error can be insignificant for a dominant cell type that makes up more than 90% of the tissue but can be quite significant for a rare cell type that is &lt; 10%, and MAE will say these are of the same importance whereas CCE would (theoretically) penalize the latter case more. </p><p> </p><p>However, I am now seeing that is makes the most sense to treat proportion estimation as a regression problem, and in that case CCE isn't an appropriate loss despite some of the good characteristics because CCE is inherently interpreting the values probability distributions. I can do some kind of optimization experiment to formally compare CCE with MAE and find out empirically which is the better performing loss for proportion estimator (but then even if we find out CCE to be the better performing one there won't be any theatrical justification for it). </p>

[the remark this answers is not in the document]
Comment (wli51) [reply examples/3.train_buddi4.ipynb:1]: <p>I did a library size normalization for the schildkraut and single cell integration. This example is based on the liver sample data that I stole from the original buddi analysis repo. I don't remember seeing library size normalization there so I also removed that step here and skipped directly to clipping. </p>

**Review by nrosed — COMMENTED** [review]
Overall I think this is a huge improvement to the code base! I just have a couple questions added within the comments. Once you address / answer / resolve them, then I will accept!

Great job!

File: examples/3.train_buddi4.ipynb (hunk 1 above)
[the remark this answers is not in the document]
Comment (nrosed) [reply examples/3.train_buddi4.ipynb:1]: no there wasn't one for this analysis,  just wanted to let you know in case you have this problem in the future, and it looks like you were already aware of it!

File: examples/1.prepare_pseudobulk.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (nrosed) [reply examples/1.prepare_pseudobulk.ipynb:1]: @falquaddoomi do you have a reccommendation here?
